zingolib/src/wallet/transaction_record.rs

```diff
@@ -0,0 +1,343 @@
+    pub total_transparent_value_spent: u64,
+
+    // All outgoing sends
+    pub outgoing_tx_data: Vec<OutgoingTxData>,
+
+    // Whether this TxID was downloaded from the server and scanned for Memos
+    pub full_tx_scanned: bool,
+
+    // Price of Zec when this Tx was created
+    pub price: Option<f64>,
+}
+
+// set
+impl TransactionRecord {
+    pub fn new(status: ConfirmationStatus, datetime: u64, transaction_id: &TxId) -> Self {
+        TransactionRecord {
+            status,
+            datetime,
+            txid: *transaction_id,
+            spent_sapling_nullifiers: vec![],
+            spent_orchard_nullifiers: vec![],
+            sapling_notes: vec![],
+            orchard_notes: vec![],
+            transparent_notes: vec![],
+            total_transparent_value_spent: 0,
+            total_sapling_value_spent: 0,
+            total_orchard_value_spent: 0,
+            outgoing_tx_data: vec![],
+            full_tx_scanned: false,
+            price: None,
+        }
+    }
+    pub fn add_spent_nullifier(&mut self, nullifier: PoolNullifier, value: u64) {
+        match nullifier {
+            PoolNullifier::Sapling(sapling_nullifier) => {
+                self.spent_sapling_nullifiers.push(sapling_nullifier);
+                self.total_sapling_value_spent += value;
+            }
+            PoolNullifier::Orchard(orchard_nullifier) => {
+                self.spent_orchard_nullifiers.push(orchard_nullifier);
+                self.total_orchard_value_spent += value;
+            }
+        }
+    }
+}
+//get
+impl TransactionRecord {
+    pub fn get_transparent_value_spent(&self) -> u64 {
+        self.total_transparent_value_spent
+    }
+    pub fn get_transaction_fee(&self) -> Result<u64, ZingoLibError> {
+        let outputted = self.value_outgoing() + self.total_change_returned();
+        if self.total_value_spent() >= outputted {
+            Ok(self.total_value_spent() - outputted)
+        } else {
+            ZingoLibError::MetadataUnderflow(format!(
+                "for txid {} with status {}: spent {}, outgoing {}, returned change {} \n {:?}",
+                self.txid,
+                self.status,
+                self.total_value_spent(),
+                self.value_outgoing(),
+                self.total_change_returned(),
+                self,
+            ))
+            .handle()
+        }
+    }
+
+    // TODO: This is incorrect in the edge case where where we have a send-to-self with
+    // no text memo and 0-value fee
+    pub fn is_outgoing_transaction(&self) -> bool {
+        (!self.outgoing_tx_data.is_empty()) || self.total_value_spent() != 0
+    }
+    pub fn is_incoming_transaction(&self) -> bool {
+        self.sapling_notes.iter().any(|note| !note.is_change())
+            || self.orchard_notes.iter().any(|note| !note.is_change())
+            || !self.transparent_notes.is_empty()
+    }
+    pub fn net_spent(&self) -> u64 {
+        assert!(self.is_outgoing_transaction());
```

Comment:
Q: What does `assert!` do in release builds?
I worry that folks may call it for incoming transactions and get away without detection.
Maybe the function should return an `Option<u64>` or perhaps (my favorite) just return 0 for incoming transactions. But maybe that's already the net effect since I suspect the two functions involved in the expression would presumably both return 0 in this case.